### DISCUSSION TITLE
How to reuse logic for every Prisma Model ? (abstract pagination ability to every models)

### DISCUSSION THREAD

**Discussion by theoludwig** [prisma/prisma#7075] [Q&A] [answered]
**TL;DR**: Is there a generic type (TypeScript) for a Prisma model, no matter what model it is, just it is a Prisma model that I can take as the parameter of my function to execute methods like `findMany` on it ?
If not how to reuse logic in Prisma ?

Hello! :wave: 

I'm new to Prisma, coming from `sequelize`/`sequelize-typescript`, I really like the way, Prisma works!

However in a typical application, it is most likely that we'll do CRUD stuff, and the "read" stuff will be paginated.
Thankfully, Prisma have a [documentation](https://www.prisma.io/docs/concepts/components/prisma-client/pagination) for that and to be honest, it is well explained, I also learned the existence of "Cursor-based pagination", I only knew "Offset pagination" before, and I felt that "Cursor-based pagination" will satisfy my needs, as I will read a lot of Data, and I'll do Infinite scrolling + real time with `socket.io`, so it's perfect for my use case.
However, I'll most likely do the same pagination system for most of my Prisma models, and I want to follow the DRY principle, I don't want to repeat over again, the logic for this pagination stuff, so here's what I did :

```ts
export interface PaginationOptions<T> {
  model: T
  query: {
    limit: number
    before?: number
    after?: number
  }
}

export const pagination = async <T>(
  options: PaginationOptions<T>
): Promise<T[]> => {
  const { model, query } = options
  // @ts-expect-error
  const items = await model.findMany({
    take: query.before != null ? query.limit * -1 : query.limit,
    skip: query.after != null || query.before != null ? 1 : undefined,
    ...(query.after != null && {
      cursor: {
        id: query.after
      }
    }),
    ...(query.before != null && {
      cursor: {
        id: query.before
      }
    })
  })
  return items
}
```

However by doing this I lose types safe ability of TypeScript, as you can I use `// @ts-expect-error` so TypeScript doesn't complain, in fact it doesn't know what is the type of the `T` generic, but that should be a Prisma Model, it doesn't matter which one but it is a Prisma model.

I searched through some Stackoverflow questions and GitHub discussions/issues, and didn't find what I needed, it was quite surprising for me, as I feel that most apps needs to reuse logic, am I missing something, is there is a better way ?

Thanks for your help! :smile:

**Comment by ryands17** [comment, marked as answer]
@Divlo 👋 

We will be introducing a CRUD wrapper that will handle the same things currently required: https://github.com/prisma/prisma/issues/5273 

It would be great if you could add the above use case in the request. As for re-using logic currently, you can do this by creating a pagination function that accepts all the pagination values and pass the `cursor` and required values. So something like this:

```ts
const users = await prisma.user.findMany({
    ...paginationParams({ before, after, limit })
})
```

**Reply by theoludwig** [reply]
I checked the issue you linked, and yep actually it is what I was searching for, but somehow I didn't find an issue for that.
I quite like the current workaround, I'll definitely test it out. :smile:

Thanks! I selected this answer. I will discuss my use case in the issue, and link to this discussion.

**Reply by platform-kit** [reply]
@ryands17 @Divlo 

I've been working on a solution to this CRUD boilerplate issue this week. What I've come up with is a Resource class which contains the reusable CRUD/BREAD operations & a simple `getModel` function which loops through the `PrismaClient`  object until it finds the model specified on the Resource's `resourceType` property.

Following that, all that is necessary in the CRUD functions is to call the corresponding methods on the model returned by `getModel()` .

Not sure if this will be of any help, but here's the code:

```ts
import { Prisma, PrismaClient } from '@prisma/client';
let Validator = require('validatorjs');
const prisma = new PrismaClient();
var fs = require('fs');
var pluralize = require('pluralize')

export class Resource {
    resourceType: string;

    constructor(type: string) {
        this.resourceType = type
    }

    getModel() {
        for (const [key, value] of Object.entries(prisma)) {
            if (typeof value == 'object' && key == this.resourceType) {
                return value
            }
        }
    }

    pluralizeType() {
        return pluralize.plural(this.resourceType)        
    }

    async browse(options?: any) {
        if (options == null) {
            options = {
                take: 10,
                skip: 0
            }
        }
        var model = await this.getModel()
        try {
            var output = await model.findMany(options)
        } catch (err) {
            output = { error: err, message: "Something went wrong." }
        }
        return output
    }

    async read(options?: any) {
        if (options == null) {
            options = {
                take: 10,
                skip: 0
            }
        }
        var model = await this.getModel()
        try {
            var output = await model.findUnique(options)
        } catch (err) {
            output = { error: err, message: "Something went wrong." }
        }
        return output
    }

    async add(data: any) {
        var apiSchema = fs.readFileSync('./prisma/api-schema/api-schema.json', { encoding: 'utf8', flag: 'r' })
        apiSchema = JSON.parse(apiSchema)
        var validationRules = apiSchema.schemas?.[this.pluralizeType()]?.['add']?.['input_validation_rules']
        console.log(validationRules)
        if (validationRules != null) {
            let validation = new Validator(data, validationRules);
            console.log(validation.passes())
            if( validation.fails()) {
                output = { validation: validation.errors, message: "Something went wrong." }
                return output
            }
        }

        var model = await this.getModel()
        var output = null
        try {
            const item = await model.create({
                data: data
            })
            output = item
        } catch (err) {
            output = { error: err, message: "Something went wrong." }
        }
        return output
    }

    async edit(id: Number, data: any) {
        var apiSchema = fs.readFileSync('./prisma/api-schema/api-schema.json', { encoding: 'utf8', flag: 'r' })
        apiSchema = JSON.parse(apiSchema)
        var validationRules = apiSchema.schemas?.[this.pluralizeType()]?.['edit']?.['input_validation_rules']
        console.log(validationRules)
        console.log(123)
        if (validationRules != null) {
            let validation = new Validator(data, validationRules);
            console.log(validation.passes())
            if( validation.fails()) {
                output = { validation: validation.errors, message: "Something went wrong." }
                return output
            }
        }
        var model = await this.getModel()
        var output = null
        try {
            const item = await model.update({
                where: {
                    id: id
                },
                data: data
            })
            output = item
        } catch (err) {
            output = { error: err, message: "Something went wrong." }
        }
        return output
    }

    async delete(id: Number) {
        var model = await this.getModel()
        var output = null
        try {
            const item = await model.delete({
                where: {
                    id: id
                }
            })
            output = item
        } catch (err) {
            output = { error: err, message: "Something went wrong." }
        }
        return output
    }


}


```

**Comment by VinayaSathyanarayana** [comment]
It will be nice to get a [OpenAPI Specification](https://github.com/OAI/OpenAPI-Specification) generated for the prisma schema file so that all of the excellent tooling available in [OpenApi Tools](https://openapi.tools/)  can be leveraged

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
